### ISSUE TITLE
Add display:table-row-group class to manage width of floated form controls

### ISSUE THREAD

**Issue by thgreasi** [twbs/bootstrap#19839]
Hi there.
I'm facing an inconsistency issue (between Firefox and Chrome) when using `float`s on a multi-line "form" that contains `input-group`s, as [demonstrated here](http://jsfiddle.net/chvg8v9z/31/).
![image](https://cloud.githubusercontent.com/assets/1295829/15010608/a4e84138-11f5-11e6-9c37-8c2f7cd96ad3.png)

A [workaround that I ended-up was to wrap every `input-group` with a `form-inline`](http://jsfiddle.net/chvg8v9z/32/), but obviously this doesn't have exactly the same result.
![image](https://cloud.githubusercontent.com/assets/1295829/15010690/090a08f4-11f6-11e6-95a8-a47cb682e1be.png)

After some hours of searching and since I couldn't find something relevant (sorry if this ends up to be a duplicate), I decided to report this here and get an official suggested way to approach this use case (which would hopefully be useful to others as well).

Tested with Bootstrap v3.3.6 on Firefox 48.0a2 and Chrome 49.

**Comment by twbs-lmvtfy** [comment]
Hi @thgreasi!

You appear to have posted a live example (http://fiddle.jshell.net/chvg8v9z/30/show/light/), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W005](https://github.com/twbs/bootlint/wiki/W005): Unable to locate jQuery, which is required for Bootstrap's JavaScript plugins to work
* line 31, column 7: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.6 ; saw what appears to be usage of Bootstrap 3.3.1
* line 35, column 7: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.6 ; saw what appears to be usage of Bootstrap 3.3.1

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by twbs-lmvtfy** [comment]
Hi @thgreasi!

You appear to have posted a live example (http://fiddle.jshell.net/chvg8v9z/29/show/light/), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* [W005](https://github.com/twbs/bootlint/wiki/W005): Unable to locate jQuery, which is required for Bootstrap's JavaScript plugins to work
* line 31, column 7: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.6 ; saw what appears to be usage of Bootstrap 3.3.1
* line 35, column 7: [W013](https://github.com/twbs/bootlint/wiki/W013): Bootstrap version might be outdated. Latest version is at least 3.3.6 ; saw what appears to be usage of Bootstrap 3.3.1

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by thgreasi** [comment]
Updated example http://jsfiddle.net/chvg8v9z/31/
Updated workaround http://jsfiddle.net/chvg8v9z/32/

**Comment by RyanZim** [comment]
Did a little cross browser testing; IE, Safari, and Opera behave the same as Chrome. Firefox is the only one that does not display `.input-group` with `width: 100%;`.

Here is a slightly simplified JSFiddle: http://jsfiddle.net/ms74k5zf/2/.

Not sure if this is a browser bug or what. :confused: 

Tested on Windows 7, Firefox 45.0.2.

**Comment by Seika85** [comment]
The problem is the styling as a table in the input-group and its children.
As a workaround for now, I'd suggest wrapping all input-groups into one input-group-wrapper and give it `display: table-row-group`.

http://jsfiddle.net/ms74k5zf/3/

I tested it in IE11, FF46 and Chrome 50. It looked exactly the same.

**Comment by RyanZim** [comment]
@Seika85 Your workaround doesn't work for me.

@cvrebert Did you file a bug with Firefox?

**Comment by thgreasi** [comment]
I can confirm that it works consistently on Firefox 47 and Chrome 50 on Android. The floated container takes only the width required to be rendered (not 100%). 
How will this be handled in future versions of bootstrap? Will the `input-group-wrapper` class be added?
Is this a bug in Firefox or in the rest browsers (since it worked the same in Firefox before and after adding the extra wrapper)?

**Comment by cvrebert** [comment]
Majority rules (basically), so prima facie, I assume this is a Firefox bug. Will try to file a bug report tomorrow.

**Comment by cvrebert** [comment]
Reduced testcase: http://jsfiddle.net/cvrebert/29vLvfLd/

**Comment by cvrebert** [comment]
Filed Firefox bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1277782

**Comment by cvrebert** [comment]
Anyway, the 2nd element floats reasonably in non-Firefox browser if you set an explicit width on the input-groups or the `.pull-left` `<div>` (and there might be further possibilities).

@thgreasi Would us adding docs about the need to set explicit widths in cases like this satisfy you sufficiently?

**Comment by thgreasi** [comment]
To be honest, I personally prefer the `.input-group-wrapper` since I need the dynamic minimum size that floats offer (and that was exactly my use case).
In my mind a floated element should behave the same no matter what it contains and that's why I thought that Firefox rendered it correctly.

On the other hand, I acknowledged that adding an extra class just for this case might be too much and that for other use cases a (min/max-)width should be enough. So, how about adding a note in the docs about the width and also link the solution posted in this issue?

**Comment by cvrebert** [comment]
@thgreasi Do you have an idea for how to implement `.input-group-wrapper` ?

**Comment by cvrebert** [comment]
(Ignore the stupid question from my deleted comment)
@thgreasi Would you be able to share what your specific, real-world form was that required such a layout?

**Comment by thgreasi** [comment]
This happened to me after migrating to bootstrap 3.
I had a floated form with some inputs and some of them had some floated action buttons right next to them, just as shown in the issue description. After combining the buttons with the inputs, the floated form went 100% width in non-firefox browsers.
The same happened when I migrated other floated input-button pairs, for example some search inputs that were placed above `<table>`s.

**Comment by cvrebert** [comment]
@thgreasi I meant, like, what are your actual form fields? Surely your actual production-deployed form has some meaningful labels and not every field is a "Search" field. What's the real thing look like?

Would using Bootstrap's grid system instead of the floats be an option for you? Why or why not?

> since I need the dynamic minimum size that floats offer (and that was exactly my use case).

Please go into more detail about why you need that and what your exact use case is.

**Comment by thgreasi** [comment]
One of the forms that drastically changed after the migrating to bootstrap 3 looks like this:
![untitled](https://cloud.githubusercontent.com/assets/1295829/15798399/64e0d95c-2a40-11e6-87bf-903dd4a86a32.png)

This was previously implemented using floats, making all rows reasonably small and placing other panels next to it whenever the screen size allowed it (:+1: to whoever defined floats). But now with bootstrap 3 I ended up using col-* along with a max-width on the container. This was actually the form that made me report this issue, since the float layout stopped working on bootstrap 3 right after I added the `input-append` for the action buttons. Floating the input and the buttons individually worked but didn't result an equal width for every row.

[This](http://pastebin.com/d9beCzSk) is how my search table page's code looks like.

I acknowledge that I could have used several other ways (closer to the bootstrap layout philosophy) for my implementation and I actually rewrote some of the pages using the grid system or min/max-widths.

The question that appeared in my mind when I faced this issue was:
> why does the `width: 100%` of the `display: table` element isn't applied based on its floated parent-element/container!?

The principle that I still have in mind (similarly to separation of concerns) is that applying styles (like floats) on any element that doesn't contain elements with strict sizing (`width:N px;`) should have the same result. This way we can divide the work, write each component having 100% width in mind and then combine all the component to create each page.
Using col-X-Y (or specific widths) on elements can indeed make the page responsive/look-good-on-many-screen-sizes but stops it from being fluid.

Thanks for all your effort so far. Keep up the great work :wink:

**Comment by twbs-lmvtfy** [comment]
Hi @thgreasi!

You appear to have posted a live example (http://pastebin.com/d9beCzSk), which is always a good first step. However, according to [the HTML5 validator](http://html5.validator.nu), **your example has some validation errors**, which might potentially be causing your issue:
* line 1, column 1 thru column 25: Start tag seen without seeing a doctype first. Expected ```` <!DOCTYPE html> ````.
* line 16, column 8 thru line 18, column 8: Misplaced non-space characters inside a table.
* line 16, column 8 thru line 18, column 8: Cannot recover after last error. Any further errors will be ignored.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by cvrebert** [comment]
Thanks for sharing those details.

CC: @mdo regarding the request for a `display:table-row-group` class

**Comment by mdo** [comment]
If this is about adding a `.d-table-row-group` class to add that style, we could add it (though that one's used far less frequently than the other `display` values). I haven't seen much in the way of folks asking for this as a utility though, so closing this as a won't fix.

Happy to revisit though if this is about a separate issue or there's more to this <3.
